Vector/ViewController/SettingsViewController.m

```diff
@@ -449,22 +448,44 @@ -(void)setNewEmailEditingEnabled:(BOOL)newEmailEditingEnabled
         // Update the flag
         _newEmailEditingEnabled = newEmailEditingEnabled;
 
-        // Update the top-rigth corner button
-        if (!_newEmailEditingEnabled)
+        if (!newEmailEditingEnabled)
         {
             newEmailTextField = nil;
-            self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemSave target:self action:@selector(onSave:)];
         }
-        else
+        
+        dispatch_async(dispatch_get_main_queue(), ^{
+            
+            // Refresh the corresponding table view cell with animation
+            [self.tableView reloadRowsAtIndexPaths:@[
+                                                     [NSIndexPath indexPathForRow:userSettingsNewEmailIndex inSection:SETTINGS_SECTION_USER_SETTINGS_INDEX]
+                                                     ]
+                                  withRowAnimation:UITableViewRowAnimationFade];
+            
+        });
+    }
+}
+
+-(void)setNewPhoneEditingEnabled:(BOOL)newPhoneEditingEnabled
+{
+    if (newPhoneEditingEnabled != _newPhoneEditingEnabled)
+    {
+        // Update the flag
+        _newPhoneEditingEnabled = newPhoneEditingEnabled;
+        
+        if (!newPhoneEditingEnabled)
         {
-            self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAdd target:self action:@selector(onAddNewEmail:)];
+            newPhoneNumberCell = nil;
         }
-
-        // And refresh the corresponding table view cell with animation
-        [self.tableView reloadRowsAtIndexPaths:@[
-                                                 [NSIndexPath indexPathForRow:userSettingsNewEmailIndex inSection:SETTINGS_SECTION_USER_SETTINGS_INDEX]
-                                                 ]
-                              withRowAnimation:UITableViewRowAnimationFade];
+        
+        dispatch_async(dispatch_get_main_queue(), ^{
```

Comment:
Because the table update failed when `newEmailEditingEnabled` and `newPhoneEditingEnabled` are changed simultaneously.